cpp/include/cudf/scalar/scalar_device_view.cuh

```diff
@@ -220,6 +220,11 @@ class fixed_point_scalar_device_view : public detail::scalar_device_view_base {
 
   __device__ void set_value(rep_type value) { *_data = value; }
 
+  /**
+   * @brief Get the value of the scalar, as a `rep_type`.
+   */
+  __device__ rep_type const& value() const noexcept { return *_data; }
```

Comment:
`value` seems like a misleading name here. It's not really the objects _value_, it's the representation as an integer that is meaningless without an associated scale.

Reply:
Agreed, but wouldn't it be dangerous to stray from the convention set in [ `cudf::fixed_point_scalar`](https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/include/cudf/scalar/scalar.hpp#L361) in `scalar.hpp`?
```c++
  /**
   * @brief Get the value of the scalar
   *
   * @param stream CUDA stream used for device memory operations.
   */
  rep_type value(rmm::cuda_stream_view stream = rmm::cuda_stream_default) const
  {
    return _data.value(stream);
  }
```
Two other related considerations:
1. Comparing `decimal32` values would be slower, because the scales would have to be normalized. In the `contains()` context, they'd've been normalized already.
2. `pair_iterator` depends on the scalar having a `value()` method.

Reply:
I think the `fixed_point_scalar::value` is similarly problematic, but I hadn't noticed it before. I'm not suggesting we always need to construct the actual value every time, my thought would be to rename this functionality to `rep()` or `representation()`.

Reply:
Actually, aren't you required to materialize the actual `decimal` type? Otherwise you could have two `decimal` values with different scales but are otherwise equivalent that will be different when you just look at the integers.

Reply:
Hmm... This will pose challenges in `cudf::detail::make_pair_iterator()`. Would this now require a specialization for `fixed_point` types, that now use `rep()` instead of `value()`? 

That doesn't seem ideal either.

Reply:
> you could have two decimal values with different scales but are otherwise equivalent

My prior reply might have clashed with this comment. 
The scales have already been compared at this stage, in `operator()()` at [Line#159](https://github.com/rapidsai/cudf/pull/7349/files#diff-d66d386d623323d882af299672d1a96abf1c8ade91dcf2cf0ac529e3b45fb329R159)
```c++
    CUDF_EXPECTS(lists.child().type() == search_key.type(),
                 "Type/Scale of search key does not match list column element type.");
```
(`data_type::operator ==()` compares scales as well.)

`contains()` requires that the scales already be normalized, prior to invocation.

Reply:
If I understand correctly, what's pending here is to figure out how to selectively use `rep()` for pair-iterators constructed from `decimal` scalars, and `value()` for all other types.

I'm leaving this conversation open until I can figure this out.

Reply:
I've fixed this now.

It turns out that simply using `value()` (or `rep()`) to return the representative value for `__device__` side comparisons via `scalar_pair_accessor` was incorrect:
```c++
  CUDA_HOST_DEVICE_CALLABLE
  const value_type operator()(size_type) const
  {
#if defined(__CUDA_ARCH__)
    return {Element(super_t::dscalar.value()), super_t::dscalar.is_valid()};
#else
    CUDF_FAIL("unsupported device scalar iterator operation");
#endif
  }
```
`scalar_pair_accessor::op()()` returns `{Element(super_t::dscalar.value())`. Note that the constructor `Element(dscalar.value())` constructs a `decimal` and uses default `0` for scale. This would have been catastrophic for decimals.

I've had to introduce an analogous `scalar_representation_pair_accessor` to return `{dscalar.rep(), dscalar.is_valid()}` instead.